precompiles/batch/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile to interact with pallet_balances instances using the ERC20 interface standard.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use evm::{ExitError, ExitReason};
+use fp_evm::{
+	Context, Log, Precompile, PrecompileFailure, PrecompileHandle, PrecompileOutput, Transfer,
+};
+use precompile_utils::{
+	keccak256, succeed, Address, Bytes, EvmDataWriter, EvmResult, FunctionModifier, LogExt,
+	LogsBuilder, PrecompileHandleExt,
+};
+use sp_core::{H160, U256};
+use sp_std::{iter::repeat, marker::PhantomData, vec, vec::Vec};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[precompile_utils::generate_function_selector]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Action {
+	BatchSome = "batchSome(address[],uint256[],bytes[],uint64[])",
+	BatchSomeUntilFailure = "batchSomeUntilFailure(address[],uint256[],bytes[],uint64[])",
+	BatchAll = "batchAll(address[],uint256[],bytes[],uint64[])",
+}
+
+pub const LOG_SUBCALL_SUCCEEDED: [u8; 32] = keccak256!("SubcallSucceeded(uint256)");
+pub const LOG_SUBCALL_FAILED: [u8; 32] = keccak256!("SubcallFailed(uint256)");
+
+pub fn log_subcall_succeeded(address: impl Into<H160>, index: usize) -> Log {
+	LogsBuilder::new(address.into()).log1(
+		LOG_SUBCALL_SUCCEEDED,
+		EvmDataWriter::new().write(U256::from(index)).build(),
+	)
+}
+
+pub fn log_subcall_failed(address: impl Into<H160>, index: usize) -> Log {
+	LogsBuilder::new(address.into()).log1(
+		LOG_SUBCALL_FAILED,
+		EvmDataWriter::new().write(U256::from(index)).build(),
+	)
+}
+
+/// Batch precompile.
+pub struct BatchPrecompile<Runtime>(PhantomData<Runtime>);
+
+impl<Runtime> Precompile for BatchPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config,
+{
+	fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let selector = handle.read_selector()?;
+
+		// No funds are transfered to the precompile address.
+		// Transfers will directly be made on the behalf of the user by the precompile.
+		handle.check_function_modifier(FunctionModifier::NonPayable)?;
+
+		Self::batch(handle, selector)
+	}
+}
+
+impl<Runtime> BatchPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config,
+{
+	fn batch(handle: &mut impl PrecompileHandle, action: Action) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let addresses: Vec<Address> = input.read()?;
+		let values: Vec<U256> = input.read()?;
+		let calls_data: Vec<Bytes> = input.read()?;
+		let gas_limits: Vec<u64> = input.read()?;
+
+		let addresses = addresses.into_iter().enumerate();
+		let values = values.into_iter().map(|x| Some(x)).chain(repeat(None));
+		let calls_data = calls_data.into_iter().map(|x| Some(x)).chain(repeat(None));
+		let gas_limits = gas_limits.into_iter().map(|x|
```

Comment:
There is a lot of array iteration here. What kind of implicit upper bound is there on these size? Should we provide an explicit one?

Reply:
You could alternatively charge for them (e.g. 2-5 gas or so per element) before any iteration.

Reply:
User pays for each byte of the input, and the input parsing is very lightweight. Not sure it is really necessary to charge more.